plugin/documents.py

```diff
@@ -606,46 +614,70 @@ def render_highlights_on_main_thread() -> None:
 
     # --- textDocument/complete ----------------------------------------------------------------------------------------
 
-    def _on_query_completions_async(self, resolve: ResolveCompletionsFn, location: int) -> None:
-        session = self.session('completionProvider', location)
-        if not session:
-            resolve([], 0)
+    def _on_query_completions_async(self, resolve_completion_list: ResolveCompletionsFn, location: int) -> None:
+        sessions = self.sessions('completionProvider')
+        if not sessions:
+            resolve_completion_list([], 0)
             return
         self.purge_changes_async()
-        can_resolve_completion_items = bool(session.get_capability('completionProvider.resolveProvider'))
-        config_name = session.config.name
-        session.send_request_async(
-            Request.complete(text_document_position_params(self.view, location), self.view),
-            lambda res: self._on_complete_result(res, resolve, can_resolve_completion_items, config_name),
-            lambda res: self._on_complete_error(res, resolve))
-
-    def _on_complete_result(self, response: Optional[Union[dict, List]], resolve: ResolveCompletionsFn,
-                            can_resolve_completion_items: bool, session_name: str) -> None:
-        response_items = []  # type: List[Dict]
-        flags = 0
+
+        completion_promises = []  # type: List[Promise[CompletionResponseWithSession]]
+        for session in sessions:
+
+            def completion_request() -> Promise:
+                return Promise(lambda resolve: session.send_request_async(
```

Comment:
Can't you use `session.send_request_task` here?

Reply:
I wasn't aware of the presence of `send_request_task` :)

But looking at the implementation of `send_request_task` I am not sure how I would pass in the session_name. I see that I can only pass in the `Request`

https://github.com/sublimelsp/LSP/pull/1582/files#diff-982c110867d9809c978a5597e4da66ce7538b327675a66b8f27214413b7c0ec1R630

Can I do something like (I haven't tried this, will try it tomorrow)
```python
        completion_promises = []  # type: List[Promise[CompletionResponseWithSession]]
        for session in sessions:

            def completion_request() -> Promise:
                return session.send_request_task(
                    Request.complete(text_document_position_params(self.view, location), self.view)
                ).then(lambda res: (res, session.config.name))

            completion_promises.append(completion_request())
```

Reply:
d19284c